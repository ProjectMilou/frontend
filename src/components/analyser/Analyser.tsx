import React from 'react';

const Analyser: React.FC = () => <div>Analyser component</div>;

export default Analyser;
