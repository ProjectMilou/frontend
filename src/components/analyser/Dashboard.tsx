import React from 'react';
import {
  LinearProgress,
  makeStyles,
  Container,
} from '@material-ui/core';
import { useTranslation } from 'react-i18next';
import * as API from '../../portfolio/APIClient';
import { ErrorCode } from '../../Errors';
import ErrorMessage from './ErrorMessage';
import PortfolioOverview from './StockListOverview';
import DashboardHeader from './DashboardHeader';

export type DashboardProps = {
  token: string;
  selectPortfolio: (id: string) => void;
};

const useStyles = makeStyles({
  createButton: {
    marginTop: '25px',
  },
  dashboard: {
    margin: '25px auto',
  },
});

const Dashboard: React.FC<DashboardProps> = ({ token, selectPortfolio }) => {
  const [portfolios, setPortfolios] = React.useState<API.PortfolioOverview[]>();
  const [error, setError] = React.useState<ErrorCode | undefined>();

  const isMounted = React.useRef(true);
  const fetch = async () => {
    setError(undefined);
    try {
      const p = await API.portfolioOverview(token);
      if (isMounted.current) {
        setPortfolios(p);
      }
    } catch (e) {
      if (isMounted.current) {
        setError(e.message);
      }
    }
  };

  React.useEffect(() => {
    fetch();
    return () => {
      isMounted.current = false;
    };
    // deps must be empty because the function should only be called on mount.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const classes = useStyles();

  return (
    <>
      <DashboardHeader />
      {!portfolios && !error && (
        <div>
          <LinearProgress color="secondary" />
        </div>
      )}
      <Container maxWidth="lg" className={classes.dashboard}>
        {error && (
          <ErrorMessage
            error={error}
            messageKey="analyser.dashboard.errorMessage"
            handling={
              error.startsWith('AUTH')
                ? {
                    buttonText: 'error.action.login',
                    action: async () => {
                      // TODO: go back to login
                    },
                  }
                : {
                    buttonText: 'error.action.retry',
                    action: fetch,
                  }
            }
          />
        )}
        {portfolios && (
          <div>
            <PortfolioOverview
              portfolios={portfolios}
              selectPortfolio={selectPortfolio}
            />
          </div>
        )}
      </Container>
    </>
  );
};

export default Dashboard;
