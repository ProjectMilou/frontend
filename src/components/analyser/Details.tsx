import React from 'react';

// props type declaration
export type DetailsProps = {
  // function to return to the dashboard
  back: () => void;
};

const Details: React.FC<DetailsProps> = () => (
  <div>
    details will be here
  </div>
);

export default Details;
