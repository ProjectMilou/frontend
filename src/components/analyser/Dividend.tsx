import React from 'react'
import { createStyles, Theme, makeStyles } from '@material-ui/core/styles';
import DividendsChart from "../shared_components/DividendsChart";
import DonutChart from '../shared_components/DonutChart';

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {
      width: '100%',
      maxWidth: 360,
      backgroundColor: theme.palette.background.paper,
    },
    indented: {
        paddingLeft: theme.spacing(7),
    },
    rowC: {
      display: "flex",
      flexDirection: "row",
      justifyContent: 'space-evenly'
    }
  }),
);

const Dividend: React.FC = () => {
    const classes = useStyles();
    return (
        <div className={classes.indented}>
            <h1>Dividends</h1>
            <div className={classes.rowC}>
                <DividendsChart/>
                <DonutChart/>
            </div>
        </div>
    )
} 
    
    
export default Dividend