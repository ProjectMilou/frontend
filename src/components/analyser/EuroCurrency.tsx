import React from 'react';
import NumberFormat from 'react-number-format';
import { makeStyles, Theme } from '@material-ui/core';

export type EuroCurrencyProps = {
  value: number;
  size?: string;
  fontWeight?: number;
  color?: string;
};

const useStyles = makeStyles<Theme, EuroCurrencyProps, string>({
  euroCurrency: {
    fontSize: (props) => props.size || '24px',
    fontWeight: (props)=> props.fontWeight || 400,
    color: (props)=> props.color || "#122654",
  },
});

const EuroCurrency: React.FC<EuroCurrencyProps> = (props) => {
  const { value } = props;
  const classes = useStyles(props);

  return (
    <NumberFormat
      value={value}
      displayType="text"
      prefix="â‚¬"
      decimalScale={2}
      fixedDecimalScale
      decimalSeparator="."
      className={classes.euroCurrency}
    />
  );
};

export default EuroCurrency;
