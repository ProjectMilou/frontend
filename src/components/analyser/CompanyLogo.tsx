import React from 'react';
import { Stock } from '../../analyser/APIClient';
import logo from '../../assets/images/placeholder.png';

type LogoProps = {
  stockOverview: Stock;
  style: string;
};

/**
 * This component displays the stock logos and replaces it with the Milou logo incase it doesn't exist
 * @param style allows the logo to be reused in different location independently styled
 */
const CompanyLogo: React.FC<LogoProps> = ({ stockOverview, style }) => {
  if (stockOverview.picture.toString() !== '') {
    return (
      <img
        className={style}
        alt="Company Logo"
        src={stockOverview.picture.toString()}
      />
    );
  }
  return <img className={style} alt="Company Logo" src={logo} />;
};

export default CompanyLogo;
