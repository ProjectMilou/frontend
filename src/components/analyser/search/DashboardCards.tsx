import React from 'react';
import { navigate } from '@reach/router';
import InfiniteScroll from 'react-infinite-scroll-component';
import {
  Card,
  CardContent,
  CardMedia,
  Grid,
  Divider,
  makeStyles,
  GridList,
  ButtonBase,
} from '@material-ui/core';
import { useTranslation } from 'react-i18next';
import * as API from '../../../analyser/APIClient';
import TextOverText from '../../shared/TextOverText';

const useStyles = makeStyles(() => ({
  card: {
    width: 289,
    margin: 10,
    transition: '0.3s',
    '&:hover': {
      boxShadow: '0 16px 70px -12.125px rgba(0,0,0,0.3)',
    },
  },
  media: {
    height: 200,
    width: 289,
    placeItems: 'center',
    alignItems: 'center',
    objectFit: 'scale-down',
    padding: 20,
  },
  content: {
    margin: 10,
  },
  divider: {
    margin: `${30}px 0`,
  },
  cardAction: {
    display: 'block',
    textAlign: 'initial',
  },
  rightBound: {
    float: 'right',
  },
  leftBound: {
    float: 'left',
  },
  paddingBottom: {
    paddingBottom: 40,
  },
  grid: {
    overflow: 'hidden' /* Hide scrollbars */,
  },
}));

export type DashboardCardsRowProps = {
  stock: API.Stock;
};

function convertPercentToColor(val: number): string {
  return val < 0 ? '#D64745' : '#50E2A8';
}

export const DashboardCardsRow: React.FC<DashboardCardsRowProps> = ({
  stock,
}) => {
  const classes = useStyles();
  const { t } = useTranslation();

  return (
    <Card className={classes.card}>
      <ButtonBase
        className={classes.cardAction}
        onClick={() => navigate(`analyser/${stock.symbol}`)}
      >
        <CardMedia
          className={classes.media}
          component="img"
          image={stock.picture.toString()}
        />
        <CardContent className={classes.content}>
          <TextOverText
            top={`${stock.symbol}`}
            bottom={`${stock.name}`}
            colorTop="#68696b"
            colorBottom="#122654"
            sizeBottom="1.3rem"
            alignment="left"
          />
          <Divider className={classes.divider} light />
          <div className={classes.paddingBottom}>
            <div className={classes.leftBound}>
              <TextOverText
                top="Last Price"
                colorTop="#68696b"
                bottom={`${stock.price}`}
                euro
                sizeBottom="1.3rem"
              />
            </div>
            <div className={classes.rightBound}>
              <TextOverText
                top={t('stock.30d')}
                bottom={`${stock.per30d.slice(0, -1)}%`}
                colorTop="#68696b"
                colorBottom={convertPercentToColor(parseFloat(stock.per30d))}
                sizeBottom="1.3rem"
              />
            </div>
          </div>
        </CardContent>
      </ButtonBase>
    </Card>
  );
};

export type DashboardCardsProps = {
  stocks: API.Stock[];
};

const DashboardCards: React.FC<DashboardCardsProps> = ({ stocks }) => {
  const classes = useStyles();

  const [items, setItems] = React.useState<API.Stock[]>(stocks.slice(0, 10));
  const [hasMore, setHasMore] = React.useState<boolean>(true);

  React.useEffect(() => {
    setItems(stocks.slice(0, 10));
    setHasMore(true);
  }, [stocks]);

  const fetchMoreData = () => {
    if (items.length >= stocks.length) {
      setHasMore(false);
      return;
    }
    setHasMore(true);
    // a fake async api call like which sends
    // 5 more stocks in 1.5 secs
    // TODO replace with async API call
    setTimeout(() => {
      const newItems = items.concat(
        stocks.slice(items.length, items.length + 5)
      );
      setItems(newItems);
    }, 1500);
  };
  return (
    <InfiniteScroll
      dataLength={items.length}
      next={fetchMoreData}
      hasMore={hasMore}
      loader={<></>}
      scrollableTarget="scrollableTable"
      endMessage={<></>}
    >
      <Grid className={classes.grid}>
        <GridList>
          {items.map((s) => (
            <DashboardCardsRow stock={s} key={s.symbol} />
          ))}
          {hasMore && <h4>Loading...</h4>}
        </GridList>
      </Grid>
    </InfiniteScroll>
  );
};

export default DashboardCards;
