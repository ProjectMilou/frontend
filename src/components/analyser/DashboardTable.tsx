import React from 'react';
import { useTranslation } from 'react-i18next';
import {
  lighten,
  ListItemText,
  makeStyles,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  Theme,
  Typography,
} from '@material-ui/core';
import classNames from 'classnames';
import * as API from '../../analyser/APIClient';
import EuroCurrency from './EuroCurrency';
import Performance from './Performance';

const useStyles = makeStyles((theme: Theme) => ({
  action: { display: 'inline-block' },
  row: {
    cursor: 'pointer',
  },
  rowHover: {
    backgroundColor: lighten(theme.palette.primary.light, 0.85),
  },
  positionCount: {
    fontSize: '24px',
  },
  disabled: {
    cursor: 'not-allowed',
  },
}));

export type DashboardTableRowProps = {
  stock: API.Stock;
};

export const DashboardTableRow: React.FC<DashboardTableRowProps> = ({
  stock
}) => {
  const [hover, setHover] = React.useState<boolean>(false);

  const { t } = useTranslation();
  const classes = useStyles();

  const selectStock = (symbol: API.Stock['symbol']) => {
    console.log(symbol)
  }

  return (
    <TableRow
      onClick={() => {
        selectStock(stock.symbol);
      }}
      onMouseEnter={() => setHover(true)}
      onMouseLeave={() => setHover(false)}
      className={classNames(classes.row, hover && classes.rowHover)}
    >
      <TableCell align="center">{stock.name}</TableCell>
      <TableCell align="center" component="th" scope="row">
        <ListItemText
          primary={<Typography color="primary">{stock.symbol}</Typography>}
          secondary={
           stock['1d'] > 0 ? t('portfolio.virtual') : t('portfolio.real')
          }
          secondaryTypographyProps={{
            color: stock['1d'] > 0 ? 'textSecondary' : 'secondary',
          }}
        />
      </TableCell>
      <TableCell align="center" className={classes.positionCount}>
        {stock.analystTargetPrice}
      </TableCell>
      <TableCell align="center">
        <EuroCurrency value={stock.valuation} />
      </TableCell>
      <TableCell align="center">
        <Performance value={stock['7d']} />
      </TableCell>
      <TableCell align="center">
        <Performance value={stock['30d']} />
      </TableCell>
    </TableRow>
  );
};

export type DashboardTableProps = {
  stocks: API.Stock[];
};

const DashboardTable: React.FC<DashboardTableProps> = ({
  stocks
}) => {
  const { t } = useTranslation();

  // TODO: Improve portfolio score visualization
  return (
    <TableContainer component={Paper}>
      <Table aria-label="simple table">
        <TableHead>
          <TableRow>
            <TableCell align="center">{t('portfolio.score')}</TableCell>
            <TableCell align="center">{t('portfolio.name')}</TableCell>
            <TableCell align="center">
              {t('portfolio.positionsCount')}
            </TableCell>
            <TableCell align="center">{t('portfolio.value')}</TableCell>
            <TableCell align="center">{t('portfolio.7d')}</TableCell>
            <TableCell align="center">{t('portfolio.1y')}</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {stocks.map((s) => (
            <DashboardTableRow
              stock={s}
              key={s.symbol}
            />
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};

export default DashboardTable;
