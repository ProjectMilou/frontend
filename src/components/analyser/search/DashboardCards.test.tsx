import { render, fireEvent } from '@testing-library/react';
import * as React from 'react';
import { createMuiTheme, ThemeProvider } from '@material-ui/core';
import DashboardCards, { DashboardCardsProps } from './DashboardCards';
import { MockOverview, MockOverviewTwo } from '../../../analyser/APIMocks';

const theme = createMuiTheme({
  typography: {
    fontFamily: [
      '-apple-system',
      'BlinkMacSystemFont',
      '"Segoe UI"',
      'Roboto',
      '"Helvetica Neue"',
      'Arial',
      'sans-serif',
      '"Apple Color Emoji"',
      '"Segoe UI Emoji"',
      '"Segoe UI Symbol"',
    ].join(','),
  },
  palette: {
    background: {
      default: '#EEF1FB',
    },
    primary: {
      light: '#415176',
      main: '#122654',
      dark: '#0c1a3a',
      contrastText: '#EEF1FB',
    },
    secondary: {
      light: '#df4f9b',
      main: '#D72483',
      dark: '#96195b',
      contrastText: '#000',
    },
    success: {
      main: '#50E2A8',
    },
    warning: {
      main: '#FFC43B',
    },
    error: {
      main: '#D64745',
    },
    lightBlue: {
      main: '#3FBCF2',
    },
    teal: {
      main: '#34CFB2',
    },
  },
});

describe('DashboardCards', () => {
  const defaultProps: DashboardCardsProps = {
    stocks: [MockOverview, MockOverviewTwo],
  };

  const renderComponent = (newProps?: Partial<DashboardCardsProps>) => {
    const props = { ...defaultProps, ...newProps };
    return {
      ...render(
        <ThemeProvider theme={theme}>
          <DashboardCards {...props} />
        </ThemeProvider>
      ),
      props,
    };
  };

  test('show empty Cards if no stocks exist', () => {
    const { container, queryByText } = renderComponent({ stocks: [] });
    // empty Cards & no error message & no table content
    expect(queryByText('analyser.dashboard.errorMessage')).toBeNull();
    expect(container.querySelector('table')).toBeNull();
    expect(container.querySelector('ul')).toBeInTheDocument();
  });

  test('display stock information', () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test('select stock on click', () => {
    const { props, getByText, queryByText } = renderComponent();
    fireEvent.click(getByText(props.stocks[0].symbol));
    expect(queryByText(props.stocks[0].symbol)).toBeInTheDocument();
    fireEvent.click(getByText(props.stocks[1].symbol));
    expect(queryByText(props.stocks[1].symbol)).toBeInTheDocument();
  });
});
