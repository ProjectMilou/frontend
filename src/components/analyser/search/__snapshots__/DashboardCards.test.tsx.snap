// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DashboardCards display stock information 1`] = `
<div>
  <div
    class="infinite-scroll-component__outerdiv"
  >
    <div
      class="infinite-scroll-component "
      style="height: auto; overflow: auto;"
    >
      <div
        class="MuiGrid-root makeStyles-grid-19"
      >
        <ul
          class="MuiGridList-root"
          style="margin: -2px;"
        >
          <div
            class="MuiPaper-root MuiCard-root makeStyles-card-11 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <button
              class="MuiButtonBase-root makeStyles-cardAction-15"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiCardMedia-root"
              >
                <img
                  alt="Company Logo"
                  class="makeStyles-media-12"
                  src="https://finnhub.io/api/logo?symbol=IBM"
                />
              </div>
              <div
                class="MuiCardContent-root makeStyles-content-13"
              >
                <div
                  class="makeStyles-compContainer-23"
                >
                  <div
                    class="makeStyles-pWrapper-24 makeStyles-pWrapper-27"
                  >
                    <p
                      class="makeStyles-top-21 makeStyles-top-25"
                    >
                      IBM
                    </p>
                  </div>
                  <div
                    class="makeStyles-pWrapper-24 makeStyles-pWrapper-27"
                  >
                    <p
                      class="makeStyles-bottom-22 makeStyles-bottom-26"
                    >
                      International Business Machines Corporation
                    </p>
                  </div>
                </div>
                <hr
                  class="MuiDivider-root makeStyles-divider-14 MuiDivider-light"
                />
                <div
                  class="makeStyles-paddingBottom-18"
                >
                  <div
                    class="makeStyles-leftBound-17"
                  >
                    <div
                      class="makeStyles-compContainer-23"
                    >
                      <div
                        class="makeStyles-pWrapper-24 makeStyles-pWrapper-30"
                      >
                        <p
                          class="makeStyles-top-21 makeStyles-top-28"
                        >
                          analyser.dashboard.view.cards.lastPrice
                        </p>
                      </div>
                      <div
                        class="makeStyles-pWrapper-24 makeStyles-pWrapper-30"
                      >
                        <p
                          class="makeStyles-bottom-22 makeStyles-bottom-29"
                        >
                          <span
                            class="makeStyles-styledNumberFormat-31 makeStyles-styledNumberFormat-33"
                          >
                            138,00 $
                          </span>
                        </p>
                      </div>
                    </div>
                  </div>
                  <div
                    class="makeStyles-rightBound-16"
                  >
                    <div
                      class="makeStyles-compContainer-23"
                    >
                      <div
                        class="makeStyles-pWrapper-24 makeStyles-pWrapper-36"
                      >
                        <p
                          class="makeStyles-top-21 makeStyles-top-34"
                        >
                          stock.30d
                        </p>
                      </div>
                      <div
                        class="makeStyles-pWrapper-24 makeStyles-pWrapper-36"
                      >
                        <p
                          class="makeStyles-bottom-22 makeStyles-bottom-35"
                        >
                          1.06%
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="MuiPaper-root MuiCard-root makeStyles-card-11 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <button
              class="MuiButtonBase-root makeStyles-cardAction-15"
              tabindex="0"
              type="button"
            >
              <div
                class="MuiCardMedia-root"
              >
                <img
                  alt="Company Logo"
                  class="makeStyles-media-12"
                  src="https://finnhub.io/api/logo?symbol=IBM"
                />
              </div>
              <div
                class="MuiCardContent-root makeStyles-content-13"
              >
                <div
                  class="makeStyles-compContainer-23"
                >
                  <div
                    class="makeStyles-pWrapper-24 makeStyles-pWrapper-39"
                  >
                    <p
                      class="makeStyles-top-21 makeStyles-top-37"
                    >
                      APPLE
                    </p>
                  </div>
                  <div
                    class="makeStyles-pWrapper-24 makeStyles-pWrapper-39"
                  >
                    <p
                      class="makeStyles-bottom-22 makeStyles-bottom-38"
                    >
                      International Business Machines Corporation
                    </p>
                  </div>
                </div>
                <hr
                  class="MuiDivider-root makeStyles-divider-14 MuiDivider-light"
                />
                <div
                  class="makeStyles-paddingBottom-18"
                >
                  <div
                    class="makeStyles-leftBound-17"
                  >
                    <div
                      class="makeStyles-compContainer-23"
                    >
                      <div
                        class="makeStyles-pWrapper-24 makeStyles-pWrapper-42"
                      >
                        <p
                          class="makeStyles-top-21 makeStyles-top-40"
                        >
                          analyser.dashboard.view.cards.lastPrice
                        </p>
                      </div>
                      <div
                        class="makeStyles-pWrapper-24 makeStyles-pWrapper-42"
                      >
                        <p
                          class="makeStyles-bottom-22 makeStyles-bottom-41"
                        >
                          <span
                            class="makeStyles-styledNumberFormat-31 makeStyles-styledNumberFormat-43"
                          >
                            138,00 $
                          </span>
                        </p>
                      </div>
                    </div>
                  </div>
                  <div
                    class="makeStyles-rightBound-16"
                  >
                    <div
                      class="makeStyles-compContainer-23"
                    >
                      <div
                        class="makeStyles-pWrapper-24 makeStyles-pWrapper-46"
                      >
                        <p
                          class="makeStyles-top-21 makeStyles-top-44"
                        >
                          stock.30d
                        </p>
                      </div>
                      <div
                        class="makeStyles-pWrapper-24 makeStyles-pWrapper-46"
                      >
                        <p
                          class="makeStyles-bottom-22 makeStyles-bottom-45"
                        >
                          1.06%
                        </p>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
          <div
            class="makeStyles-loading-20"
            style="width: 50%; height: 184px; padding: 2px;"
          >
            <div
              class="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
              role="progressbar"
              style="width: 40px; height: 40px;"
            >
              <svg
                class="MuiCircularProgress-svg"
                viewBox="22 22 44 44"
              >
                <circle
                  class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
                  cx="44"
                  cy="44"
                  fill="none"
                  r="20.2"
                  stroke-width="3.6"
                />
              </svg>
            </div>
          </div>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
